File: dynamic-swap-program/prover/src/escrow_open.rs
use std::collections::HashMap;

use crate::{
    bytes_to_decimal_string,
    ffi::{self, CircuitId},
    proof::{negate_and_compress_proof, OrderProof, ProofError},
    utxo::utxo_witness_entries,
    ProofInputUtxo,
};

/// Proof inputs for the `escrow_open` circuit (`create_escrow`): 2-in (source,
/// maker_funding) / 3-out (order, reservation, maker_change), the exact supported
/// IN2_OUT3 shape with no padding. No source change output: the source UTXO must
/// match `order_amount` exactly. `order_amount` is the one private witness shared
/// across the order UTXO, the reservation size (`order_amount * max_price`), and
/// the maker-change decrement.
#[derive(Debug, Clone)]
pub struct EscrowOpenProofInputs {
    pub public_input_hash: [u8; 32],
    pub private_tx_hash: [u8; 32],
    pub max_price: u64,
    pub created_at: u64,
    /// The escrow_authority PDA's owner-hash (`EscrowAuthorityOwnerHash`),
    /// bound to `OrderOut.Owner`.
    pub escrow_authority_owner_hash: [u8; 32],
    /// The pair's source-asset commitment (`SourceAsset`), bound to
    /// `SourceIn.Asset`.
    pub source_asset: [u8; 32],
    /// The pair's destination-asset commitment (`DestinationAsset`), bound to
    /// `MakerFunding.Asset`.
    pub destination_asset: [u8; 32],
    pub order_amount: u64,
    pub source_in: ProofInputUtxo,
    pub maker_funding: ProofInputUtxo,
    pub order_out: ProofInputUtxo,
    pub reservation_out: ProofInputUtxo,
    pub maker_change: ProofInputUtxo,
    pub external_data_hash: [u8; 32],
    pub private_tx_blinding: [u8; 32],
}

impl EscrowOpenProofInputs {
    fn witness(&self) -> ffi::WitnessMap {
        let mut map = HashMap::new();
        map.insert(
            "Public_PublicInputHash".to_string(),
            vec![bytes_to_decimal_string(&self.public_input_hash)],
        );
        map.insert(
            "Public_PrivateTxHash".to_string(),
            vec![bytes_to_decimal_string(&self.private_tx_hash)],
        );
        map.insert("MaxPrice".to_string(), vec![self.max_price.to_string()]);
        map.insert(
            "Public_CreatedAt".to_string(),
            vec![self.created_at.to_string()],
        );
        map.insert(
            "Public_EscrowAuthorityOwnerHash".to_string(),
            vec![bytes_to_decimal_string(&self.escrow_authority_owner_hash)],
        );
        map.insert(
            "Public_SourceAsset".to_string(),
            vec![bytes_to_decimal_string(&self.source_asset)],
        );
        map.insert(
            "Public_DestinationAsset".to_string(),
            vec![bytes_to_decimal_string(&self.destination_asset)],
        );
        map.insert(
            "OrderAmount".to_string(),
            vec![self.order_amount.to_string()],
        );
        map.insert(
            "ExternalDataHash".to_string(),
            vec![bytes_to_decimal_string(&self.external_data_hash)],
        );
        map.insert(
            "PrivateTxBlinding".to_string(),
            vec![bytes_to_decimal_string(&self.private_tx_blinding)],
        );
        for (key, value) in utxo_witness_entries(&self.source_in, "SourceIn")
            .into_iter()
            .chain(utxo_witness_entries(&self.maker_funding, "MakerFunding"))
            .chain(utxo_witness_entries(&self.order_out, "OrderOut"))
            .chain(utxo_witness_entries(
                &self.reservation_out,
                "ReservationOut",
            ))
            .chain(utxo_witness_entries(&self.maker_change, "MakerChange"))
        {
            map.insert(key, value);
        }
        map
    }

    pub fn prove(&self) -> Result<OrderProof, ProofError> {
        negate_and_compress_proof(&ffi::prove(CircuitId::EscrowOpen, &self.witness())?)
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use super::*;
    use crate::utxo::expected_utxo_witness_keys;

    fn sample() -> EscrowOpenProofInputs {
        EscrowOpenProofInputs {
            public_input_hash: [1; 32],
            private_tx_hash: [2; 32],
            max_price: 100,
            created_at: 1_700_000_000,
            escrow_authority_owner_hash: [6; 32],
            source_asset: [7; 32],
            destination_asset: [8; 32],
            order_amount: 50,
            source_in: ProofInputUtxo::default(),
            maker_funding: ProofInputUtxo::default(),
            order_out: ProofInputUtxo::default(),
            reservation_out: ProofInputUtxo::default(),
            maker_change: ProofInputUtxo::default(),
            external_data_hash: [5; 32],
            private_tx_blinding: [9; 32],
        }
    }

    #[test]
    fn witness_key_set_matches_circuit_fields() {
        let witness = sample().witness();
        let keys: HashSet<&str> = witness.keys().map(String::as_str).collect();

        let mut expected: Vec<String> = vec![
            "Public_PublicInputHash".to_string(),
            "Public_PrivateTxHash".to_string(),
            "MaxPrice".to_string(),
            "Public_CreatedAt".to_string(),
            "Public_EscrowAuthorityOwnerHash".to_string(),
            "Public_SourceAsset".to_string(),
            "Public_DestinationAsset".to_string(),
            "OrderAmount".to_string(),
            "ExternalDataHash".to_string(),
            "PrivateTxBlinding".to_string(),
        ];
        for prefix in [
            "SourceIn",
            "MakerFunding",
            "OrderOut",
            "ReservationOut",
            "MakerChange",
        ] {
            expected.extend(expected_utxo_witness_keys(prefix));
        }

        let expected: HashSet<&str> = expected.iter().map(String::as_str).collect();
        assert_eq!(keys, expected);
    }
}

File: dynamic-swap-program/prover/circuits/escrow_open/escrow_open.go
package escrow_open

import (
	"github.com/consensys/gnark/frontend"

	"zolana/prover/circuits/gadget"
	spp "zolana/prover/circuits/spp_transaction/shared"
)

// Circuit binds create_escrow's 2-in/3-out real shape (taker's source UTXO +
// maker's funding UTXO spent; order, reservation, and maker-change UTXOs
// created), the exact supported IN2_OUT3 shape with no padding on either side.
// The taker signs SourceIn; the program signs the escrow-authority-owned
// MakerFunding input, which authorizes the data-bearing program outputs. Requires
// the source input to match OrderAmount exactly (no taker change output):
// create_escrow's instruction data already sits at Solana's whole-transaction
// size limit, so each side pre-consolidates its own note off the create path.
// OrderAmount is the one private witness shared across the order UTXO's amount,
// the reservation's worst-case size (order_amount * max_price), and the
// maker-change decrement.
type Circuit struct {
	Public PublicInputs

	// Each UTXO carries the raw id of the tree it lives in as a sibling witness;
	// spp.UtxoHashCircuit folds it in as the second Poseidon element.
	SourceIn           spp.UtxoCircuitFields
	SourceInTreeID     frontend.Variable
	MakerFunding       spp.UtxoCircuitFields
	MakerFundingTreeID frontend.Variable

	OrderOut             spp.UtxoCircuitFields
	OrderOutTreeID       frontend.Variable
	ReservationOut       spp.UtxoCircuitFields
	ReservationOutTreeID frontend.Variable
	MakerChange          spp.UtxoCircuitFields
	MakerChangeTreeID    frontend.Variable

	OrderAmount frontend.Variable

	// MaxPrice is a private witness (not a public input): it is committed into
	// OrderOut.DataHash and later re-opened by escrow_settle, but never revealed
	// on-chain -- keeping it private is what hides the eventual settle-vs-refund
	// outcome.
	MaxPrice frontend.Variable

	ExternalDataHash  frontend.Variable
	PrivateTxBlinding frontend.Variable
}

func (c *Circuit) Define(api frontend.API) error {
	api.AssertIsDifferent(c.OrderAmount, 0)

	// MaxPrice is a free private witness, so pin it to 64 bits. Without this the
	// prover could pick a field-sized MaxPrice whose product OrderAmount*MaxPrice
	// still reduces to a valid 64-bit reservation, then commit that garbage value
	// into OrderOut.DataHash -- where escrow_settle re-opens it and feeds it to a
	// bounded comparator (ExecutionPrice <= MaxPrice), whose result is undefined
	// on an out-of-range operand. Bounding it here (the value's origin) makes it
	// 64-bit everywhere it is later re-opened from the same DataHash.
	api.ToBinary(c.MaxPrice, 64)

	sourceInHash := c.checkSourceInputUtxo(api)
	makerFundingHash := c.checkMakerFundingInputUtxo(api)

	orderOutHash := c.checkOrderOutputUtxo(api)
	reservationOutHash := c.checkReservationOutputUtxo(api, orderOutHash)
	makerChangeHash := c.checkMakerChangeOutputUtxo(api)

	privateTxHashInputs{
		SourceInputUtxoHash:       sourceInHash,
		MakerFundingInputUtxoHash: makerFundingHash,
		OrderOutputUtxoHash:       orderOutHash,
		ReservationOutputUtxoHash: reservationOutHash,
		MakerChangeOutputUtxoHash: makerChangeHash,
		ExternalDataHash:          c.ExternalDataHash,
		PrivateTxBlinding:         c.PrivateTxBlinding,
		PrivateTxHash:             c.Public.PrivateTxHash,
	}.Check(api)

	c.Public.Check(api)
	return nil
}

// PublicInputs folds PrivateTxHash with the escrow term visible to the program
// (CreatedAt), the escrow_authority owner-hash, and the two asset bindings
// (SourceAsset, DestinationAsset). MaxPrice is a private witness, committed only
// into OrderOut.DataHash, which is what hides the eventual settle-vs-refund
// outcome. The recipient is NOT here either: it is bound in-circuit to
// SourceIn.Owner (the taker whose funds are escrowed) and committed only into
// OrderOut.DataHash, so it stays as confidential as MaxPrice -- see
// checkOrderOutputUtxo.
type PublicInputs struct {
	PublicInputHash frontend.Variable `gnark:",public"`

	PrivateTxHash frontend.Variable
	CreatedAt     frontend.Variable
	// The escrow_authority PDA's owner-hash, recomputed on-chain by the native
	// program and bound to OrderOut.Owner (and thus ReservationOut.Owner). Without
	// it OrderOut.Owner is a free witness and a caller could mint the escrow/
	// reservation UTXOs to an owner it controls, then spend the maker-funded
	// reservation directly.
	EscrowAuthorityOwnerHash frontend.Variable
	// The pair's source asset, fed on-chain from Pair.source_asset and bound to
	// SourceIn.Asset. Without it a caller could escrow a worthless token and
	// extract the destination asset on settle.
	SourceAsset frontend.Variable
	// The pair's destination asset, fed on-chain from Pair.destination_asset and
	// bound to MakerFunding.Asset (and thus ReservationOut/MakerChange). Without it
	// the maker could fund with a worthless token that the taker would be paid on
	// settle.
	DestinationAsset frontend.Variable
}

func (p PublicInputs) Check(api frontend.API) {
	publicInputHash := gadget.PoseidonHash(api, []frontend.Variable{
		p.PrivateTxHash,
		p.CreatedAt,
		p.EscrowAuthorityOwnerHash,
		p.SourceAsset,
		p.DestinationAsset,
	})
	api.AssertIsEqual(p.PublicInputHash, publicInputHash)
}

type privateTxHashInputs struct {
	SourceInputUtxoHash       frontend.Variable
	MakerFundingInputUtxoHash frontend.Variable
	OrderOutputUtxoHash       frontend.Variable
	ReservationOutputUtxoHash frontend.Variable
	MakerChangeOutputUtxoHash frontend.Variable
	ExternalDataHash          frontend.Variable
	PrivateTxBlinding         frontend.Variable
	PrivateTxHash             frontend.Variable
}

func (t privateTxHashInputs) Check(api frontend.API) {
	// The real shape is 2-in/3-out, exactly the supported IN2_OUT3 shape --
	// no padding needed on either side. Output order (order, reservation,
	// maker_change) must match the native program's output indices and the SDK.
	inputHashes := []frontend.Variable{
		t.SourceInputUtxoHash,
		t.MakerFundingInputUtxoHash,
	}
	outputHashes := []frontend.Variable{
		t.OrderOutputUtxoHash,
		t.ReservationOutputUtxoHash,
		t.MakerChangeOutputUtxoHash,
	}
	addressHashes := []frontend.Variable{
		frontend.Variable(0),
		frontend.Variable(0),
	}

	privateTxHash := spp.PrivateTxHashCircuit(
		api,
		inputHashes,
		outputHashes,
		addressHashes,
		t.ExternalDataHash,
		t.PrivateTxBlinding,
	)
	api.AssertIsEqual(privateTxHash, t.PrivateTxHash)
}

func (c *Circuit) checkSourceInputUtxo(api frontend.API) frontend.Variable {
	api.AssertIsEqual(c.SourceIn.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.SourceIn.RingDataHash, 0)
	api.AssertIsEqual(c.SourceIn.RingProgramID, 0)
	api.AssertIsEqual(c.SourceIn.DataHash, 0)
	// Bind the escrowed asset to the pair's source asset so a worthless token
	// cannot stand in for it.
	api.AssertIsEqual(c.SourceIn.Asset, c.Public.SourceAsset)
	// No change output: the source input must be exactly OrderAmount.
	api.AssertIsEqual(c.SourceIn.Amount, c.OrderAmount)
	return spp.UtxoHashCircuit(api, c.SourceIn, c.SourceInTreeID)
}

func (c *Circuit) checkMakerFundingInputUtxo(api frontend.API) frontend.Variable {
	api.AssertIsEqual(c.MakerFunding.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.MakerFunding.RingDataHash, 0)
	api.AssertIsEqual(c.MakerFunding.RingProgramID, 0)
	api.AssertIsEqual(c.MakerFunding.DataHash, 0)
	// Bind the maker's funding asset to the pair's destination asset so the maker
	// cannot fund with a worthless token that the taker would be paid on settle.
	api.AssertIsEqual(c.MakerFunding.Asset, c.Public.DestinationAsset)
	return spp.UtxoHashCircuit(api, c.MakerFunding, c.MakerFundingTreeID)
}

// checkOrderOutputUtxo commits (recipient, MaxPrice, CreatedAt) into the order
// UTXO's DataHash so settle can later re-derive the same escrow terms from the
// UTXO alone. The recipient is bound to SourceIn.Owner: the escrow's payout on
// settle/refund goes to the same party whose source funds are being escrowed (the
// taker), enforced in-circuit rather than trusted from a caller-supplied field --
// SourceIn.Owner is pinned to the real spent source leaf via SourceInputUtxoHash
// -> PrivateTxHash. OrderOut.Owner is bound to the public EscrowAuthorityOwnerHash
// (and ReservationOut.Owner is tied to it in checkReservationOutputUtxo), so both
// order and reservation UTXOs are provably owned by the pair's escrow_authority
// PDA -- only the native program can spend them, via settle.
func (c *Circuit) checkOrderOutputUtxo(api frontend.API) frontend.Variable {
	api.AssertIsEqual(c.OrderOut.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.OrderOut.RingDataHash, 0)
	api.AssertIsEqual(c.OrderOut.RingProgramID, 0)
	api.AssertIsEqual(c.OrderOut.Owner, c.Public.EscrowAuthorityOwnerHash)
	api.AssertIsEqual(c.OrderOut.Asset, c.SourceIn.Asset)
	api.AssertIsEqual(c.OrderOut.Amount, c.OrderAmount)
	api.AssertIsEqual(c.OrderOut.DataHash, gadget.PoseidonHash(api, []frontend.Variable{
		c.SourceIn.Owner,
		c.MaxPrice,
		c.Public.CreatedAt,
	}))
	return spp.UtxoHashCircuit(api, c.OrderOut, c.OrderOutTreeID)
}

// checkReservationOutputUtxo binds the reservation's DataHash to the order UTXO's
// own hash, so settle can prove in-circuit that the reservation UTXO it spends
// really belongs to this specific order. The reservation is the maker's worst-case
// liquidity (order_amount * max_price) in the destination asset, escrowed under the
// escrow_authority PDA.
func (c *Circuit) checkReservationOutputUtxo(api frontend.API, orderOutHash frontend.Variable) frontend.Variable {
	api.AssertIsEqual(c.ReservationOut.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.ReservationOut.RingDataHash, 0)
	api.AssertIsEqual(c.ReservationOut.RingProgramID, 0)
	api.AssertIsEqual(c.ReservationOut.Asset, c.MakerFunding.Asset)
	api.AssertIsEqual(c.ReservationOut.Owner, c.OrderOut.Owner)
	api.AssertIsEqual(c.ReservationOut.DataHash, orderOutHash)

	api.AssertIsEqual(c.ReservationOut.Amount, api.Mul(c.OrderAmount, c.MaxPrice))

	return spp.UtxoHashCircuit(api, c.ReservationOut, c.ReservationOutTreeID)
}

// checkMakerChangeOutputUtxo returns the maker's unspent funding (funding -
// reserved) to the maker's own note -- same asset and owner as MakerFunding, so a
// co-signing taker cannot redirect it. The 64-bit range check rejects an
// over-reservation (funding < reserved would wrap the field subtraction).
func (c *Circuit) checkMakerChangeOutputUtxo(api frontend.API) frontend.Variable {
	api.AssertIsEqual(c.MakerChange.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.MakerChange.RingDataHash, 0)
	api.AssertIsEqual(c.MakerChange.RingProgramID, 0)
	api.AssertIsEqual(c.MakerChange.DataHash, 0)
	api.AssertIsEqual(c.MakerChange.Asset, c.MakerFunding.Asset)
	api.AssertIsEqual(c.MakerChange.Owner, c.MakerFunding.Owner)

	reserved := api.Mul(c.OrderAmount, c.MaxPrice)
	api.AssertIsEqual(c.MakerChange.Amount, api.Sub(c.MakerFunding.Amount, reserved))
	api.ToBinary(c.MakerChange.Amount, 64)

	return spp.UtxoHashCircuit(api, c.MakerChange, c.MakerChangeTreeID)
}
